File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/complex-numbers/complex-numbers.js
export class ComplexNumber {
  constructor(real, imaginary) {
    this._real = real;
    this._imaginary = imaginary;
  }

  get real() {
    return this._real;
  }

  get imag() {
    return this._imaginary;
  }

  add(other) {
    return new ComplexNumber(
      this.real + other.real,
      this.imag + other.imag
    );
  }

  sub(other) {
    return new ComplexNumber(
      this.real - other.real,
      this.imag - other.imag
    );
  }

  mul(other) {
    return new ComplexNumber(
      this.real * other.real - this.imag * other.imag,
      this.imag * other.real + this.real * other.imag
    );
  }

  div(other) {
    const denominator = other.real ** 2 + other.imag ** 2;
    return new ComplexNumber(
      (this.real * other.real + this.imag * other.imag) / denominator,
      (this.imag * other.real - this.real * other.imag) / denominator
    );
  }

  get abs() {
    return Math.sqrt(this.real ** 2 + this.imag ** 2);
  }

  get conj() {
    return new ComplexNumber(this._real, -this._imaginary);
  }

  get exp() {
    const eA = Math.exp(this.real);
    return new ComplexNumber(
      eA * Math.cos(this.imag),
      eA * Math.sin(this.imag)
    );
  }
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/cpp/exercises/practice/complex-numbers/complex_numbers.cpp
#include "complex_numbers.h"

namespace complex_numbers {

Complex::Complex(double real, double imaginary) 
    : real_(real), imaginary_(imaginary) {}

double Complex::real() const {
    return real_;
}

double Complex::imag() const {
    return imaginary_;
}

Complex Complex::conj() const {
    return Complex(real_, -imaginary_);
}

double Complex::abs() const {
    return std::sqrt(real_ * real_ + imaginary_ * imaginary_);
}

Complex Complex::operator+(const Complex& other) const {
    return Complex(real_ + other.real_, imaginary_ + other.imaginary_);
}

Complex Complex::operator-(const Complex& other) const {
    return Complex(real_ - other.real_, imaginary_ - other.imaginary_);
}

Complex Complex::operator*(const Complex& other) const {
    return Complex(
        real_ * other.real_ - imaginary_ * other.imaginary_,
        imaginary_ * other.real_ + real_ * other.imaginary_
    );
}

Complex Complex::operator/(const Complex& other) const {
    double denominator = other.real_ * other.real_ + other.imaginary_ * other.imaginary_;
    return Complex(
        (real_ * other.real_ + imaginary_ * other.imaginary_) / denominator,
        (imaginary_ * other.real_ - real_ * other.imaginary_) / denominator
    );
}

Complex Complex::exp() const {
    double exp_real = std::exp(real_);
    return Complex(
        exp_real * std::cos(imaginary_),
        exp_real * std::sin(imaginary_)
    );
}

// Non-member operator implementations
Complex operator+(double lhs, const Complex& rhs) {
    return Complex(lhs + rhs.real(), rhs.imag());
}

Complex operator-(double lhs, const Complex& rhs) {
    return Complex(lhs - rhs.real(), -rhs.imag());
}

Complex operator*(double lhs, const Complex& rhs) {
    return Complex(lhs * rhs.real(), lhs * rhs.imag());
}

Complex operator/(double lhs, const Complex& rhs) {
    double denominator = rhs.real() * rhs.real() + rhs.imag() * rhs.imag();
    return Complex(
        (lhs * rhs.real()) / denominator,
        (-lhs * rhs.imag()) / denominator
    );
}

}  // namespace complex_numbers
